pallets/loans/src/types.rs

```diff
@@ -196,29 +320,34 @@ where
 				debt.checked_sub(&write_off_amount)?
 			}
 		};
-		match self.loan_type {
-			LoanType::BulletLoan(bl) => {
-				bl.present_value(debt, self.origination_date, now, self.interest_rate_per_sec)
-			}
-			LoanType::CreditLine(cl) => cl.present_value(debt),
-			LoanType::CreditLineWithMaturity(clm) => {
-				clm.present_value(debt, self.origination_date, now, self.interest_rate_per_sec)
-			}
+
+		match self.pricing.valuation_method {
+			ValuationMethod::DiscountedCashFlows(bl) => bl.present_value(
+				debt,
+				self.loan.schedule.maturity_date(),
+				self.origination_date,
+				now,
+				self.pricing.interest_rate_per_sec,
+			),
+			ValuationMethod::OutstandingDebt(cl) => cl.present_value(debt),
 		}
 	}
 
-	pub fn max_borrow_amount(&self, debt: Balance) -> Balance {
-		match self.loan_type {
-			LoanType::BulletLoan(bl) => bl.max_borrow_amount(self.total_borrowed),
-			LoanType::CreditLine(cl) => cl.max_borrow_amount(debt),
-			LoanType::CreditLineWithMaturity(clm) => clm.max_borrow_amount(debt),
+	pub fn max_borrow_amount(&self, debt: Balance) -> Result<Balance, ArithmeticError> {
+		Ok(match self.pricing.restrictions.max_borrow_amount {
+			MaxBorrowAmount::UpToTotalBorrowed { advance_rate } => advance_rate
+				.ensure_mul_int(self.pricing.collateral_value)?
+				.checked_sub(&self.total_borrowed),
+			MaxBorrowAmount::UpToOutstandingDebt { advance_rate } => advance_rate
+				.ensure_mul_int(self.pricing.collateral_value)?
+				.checked_sub(&debt),
```

Comment:
We are currently saturating to 0 if this fails, but should we? I mean, if the `debt` is > `max_borrow_amount` should be an error instead?

Reply:
No that should be possible. If a max borrowed amount is 100 and someone borrows 100, then immediately 1 block later `debt > max_borrow_amount`, but this is expected. They just won't be able to borrow more, but this is not an error case.